common/src/main/java/org/conscrypt/ConscryptEngine.java

```diff
@@ -1372,14 +1372,17 @@ public void onSSLStateChange(int type, int val) {
 
     @Override
     public boolean onNewSessionEstablished(long sslSessionNativePtr) {
+        SslSessionWrapper sessionWrapper;
         try {
-            // Cache the newly established session.
-            AbstractSessionContext ctx = sessionContext();
-            ctx.cacheSession(SslSessionWrapper.newInstance(sslSessionNativePtr, sslSession));
-            return true;
-        } catch (Exception ignored) {
+            sessionWrapper = SslSessionWrapper.newInstance(sslSessionNativePtr, sslSession);
+        } catch (Throwable e) {
```

Comment:
Much clearer. Thanks! One small comment: you should be able to avoid checking for `NativeRef.SSL_SESSION` throwing. I suppose it does technically throw NPE, but then all this freeing is moot anyway. (Also, you'd have crashed before that point in up_ref and the callback will never give you a NULL pointer.)